### PR TITLE
Add mallinfo.h, implement it for dlmalloc, and convert testcases

### PR DESCRIPTION
**Pull request by achamayou**
Implements #3410, in order to solve #3262.

The implementation for snmalloc will follow in a separate PR, since it requires an update of the submodule.

### CODE REVIEW

File: 3rdparty/dlmalloc/allocator.c [hunk 1]
```diff
@@ -115,3 +116,18 @@ size_t oe_allocator_malloc_usable_size(void* ptr)
 {
     return dlmalloc_usable_size(ptr);
 }
+
+int oe_allocator_mallinfo(oe_mallinfo_t* info)
+{
+    info->max_total_heap_size = _heap_end - _heap_start;
```
Comment (anakrish) [inline 3rdparty/dlmalloc/allocator.c:122]: Can _heap_end ever decrease? If so, we'd have to maintain a running max heap size.

**Review by anakrish — APPROVED** [review]

File: include/openenclave/advanced/mallinfo.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Obtain current memory usage statistics.
+ *
+ * Users may make this call on allocators that support it, to find out how much
+ * memory can be allocated in total, how much is allocated at the moment, and
+ * the high watermark of allocation so far.
+ *
+ * @param[out] info An oe_mallinfo_t struct, to be populated by the allocator.
+ *
+ * @returns 0 on success.
```
Comment (dthaler) [inline include/openenclave/advanced/mallinfo.h:45]: What does it return if the allocator doesn't support it?

File: tests/bigmalloc/enc/enc.c [hunk 3]
```diff
@@ -14,23 +15,19 @@ oe_result_t test_malloc()
...
 
     /* Determine how much heap memory remains */
     {
-        const uint8_t* base = (const uint8_t*)__oe_get_heap_base();
-        const uint8_t* brk = (const uint8_t*)dlmalloc_sbrk(0);
-        const uint8_t* end = (const uint8_t*)__oe_get_heap_end();
-
-        /* Sanity checks */
-        if (!(base <= brk && brk < end))
+        static oe_mallinfo_t info;
```
Comment (dthaler) [inline tests/bigmalloc/enc/enc.c:22]: Why does this need to be static?

File: tests/memory/enc/fragment.c [hunk 4]
```diff
@@ -16,27 +17,12 @@
...
-};
-
-struct mallinfo dlmallinfo(void);
-
 static size_t _get_heap_size()
 {
-    // call malloc to update the global data.
-    free(malloc(1));
-    return dlmallinfo().keepcost;
+    static oe_mallinfo_t info;
```
Comment (dthaler) [inline tests/memory/enc/fragment.c:22]: Why static?

File: 3rdparty/dlmalloc/allocator.c (hunk 1 above)
Comment (achamayou) [reply 3rdparty/dlmalloc/allocator.c:122]: As suggested on #3484, not currently, no.

File: include/openenclave/advanced/mallinfo.h (hunk 2 above)
Comment (achamayou) [reply include/openenclave/advanced/mallinfo.h:45]: Answered on #3484

File: tests/bigmalloc/enc/enc.c (hunk 3 above)
Comment (achamayou) [reply tests/bigmalloc/enc/enc.c:22]: No good reason, will fix.

File: tests/memory/enc/fragment.c (hunk 4 above)
Comment (achamayou) [reply tests/memory/enc/fragment.c:22]: No good reason, will fix.

File: 3rdparty/dlmalloc/allocator.c (hunk 1 above)
Comment (achamayou) [reply 3rdparty/dlmalloc/allocator.c]: I've split `_max_total_heap` for clarity, as suggested by @mjp41 on the other PR.

**Comment by achamayou** [comment]
bors try

File: include/openenclave/advanced/mallinfo.h (hunk 2 above)
Comment (achamayou) [reply include/openenclave/advanced/mallinfo.h]: Converted to oe_result_t, as discussed on #3484.

**Comment by achamayou** [comment]
bors try

**Comment by achamayou** [comment]
Closing in favour of #3484

**Pull request closed without merging**